File: asf/analyzer.py
from __future__ import annotations

from pathlib import Path
from typing import Optional, Sequence

from asf.assumption import AssumptionEngine
from asf.config import ASFConfig
from asf.confidence import ConfidenceEngine
from asf.db.database import Database
from asf.evidence import EvidenceLoader, EvidenceMapper
from asf.extraction import ClaimExtractor
from asf.gaps import GapEngine
from asf.graph import GraphModel
from asf.ingestion import IngestionPipeline
from asf.models import (
    AnalysisResult,
    Assumption,
    Claim,
    Evidence,
    VerificationResult,
)
from asf.verification import VerificationEngine


class Analyzer:
    def __init__(self, config: ASFConfig | None = None):
        self.config = config or ASFConfig.default()
        self.pipeline = IngestionPipeline()
        self.claim_extractor = ClaimExtractor()
        self.assumption_engine = AssumptionEngine()
        self.evidence_loader = EvidenceLoader(
            field_mappings=self.config.field_mappings or None
        )
        self.evidence_mapper = EvidenceMapper()
        self.verification_engine = VerificationEngine()
        self.confidence_engine = ConfidenceEngine()
        self.gap_engine = GapEngine()
        self.graph_model = GraphModel()
        self.result = AnalysisResult()
        self.db: Optional[Database] = None

        if self.config.db_path:
            self.db = Database(self.config.db_path)

    def analyze(
        self,
        document_paths: list[str | Path],
        evidence_paths: list[str | Path] | None = None,
        persist: bool = False,
    ) -> AnalysisResult:
        self.result = AnalysisResult()

        claims = self._process_documents(document_paths)
        assumptions = self.assumption_engine.convert_many(claims)

        evidence_records = []
        if evidence_paths:
            evidence_records = self._load_evidence(evidence_paths)

        evidence_list = list(evidence_records)

        verifications = []
        for assumption in assumptions:
            matching_evidence = self._find_matching_evidence(assumption, evidence_list)
            verification = self.verification_engine.verify(assumption, matching_evidence)
            verification.confidence = self.confidence_engine.compute_verification_confidence(
                verification, matching_evidence
            )
            verifications.append(verification)

            assumption.confidence = self.confidence_engine.compute_assumption_confidence([verification])
            if verification.result == VerificationResult.VERIFIED:
                from asf.models import VerificationStatus
                assumption.verification_status = VerificationStatus.VERIFIED
            elif verification.result == VerificationResult.CONTRADICTED:
                from asf.models import VerificationStatus
                assumption.verification_status = VerificationStatus.CONTRADICTED
            elif verification.result == VerificationResult.PARTIALLY_VERIFIED:
                from asf.models import VerificationStatus
                assumption.verification_status = VerificationStatus.IN_REVIEW

        gaps = self.gap_engine.generate_gaps(assumptions, verifications)

        self.result = AnalysisResult(
            claims=claims,
            assumptions=assumptions,
            evidence=evidence_list,
            verifications=verifications,
            gaps=gaps,
        )

        self.graph_model.build(self.result)

        if persist and self.db:
            self._persist()

        return self.result

    def _process_documents(self, paths: list[str | Path]) -> list[Claim]:
        all_claims: list[Claim] = []
        for path in paths:
            path = Path(path)
            if not path.exists():
                continue
            text = self.pipeline.parse_text(path)
            doc_meta = self.pipeline.get_document_metadata(path)
            claims = self.claim_extractor.extract(
                text,
                source_document=doc_meta["filename"],
                source_location=str(path),
            )
            all_claims.extend(claims)
        return all_claims

    def _load_evidence(self, paths: list[str | Path]) -> list[Evidence]:
        evidence_list: list[Evidence] = []
        for path in paths:
            path = Path(path)
            if not path.exists():
                continue
            evidence = self.evidence_loader.load(path, auto_map=True)
            evidence_list.append(evidence)
        return evidence_list

    def _find_matching_evidence(self, assumption: Assumption, evidence_list: list[Evidence]) -> list[Evidence]:
        compatible_types = self.evidence_mapper.get_compatible_source_types(assumption.assumption_type)
        if not compatible_types:
            return evidence_list
        matched = [e for e in evidence_list if e.source_type in compatible_types]
        if not matched:
            return evidence_list
        return matched

    def _persist(self) -> None:
        if not self.db:
            return
        for claim in self.result.claims:
            self.db.insert_claim(claim.model_dump())
        for assumption in self.result.assumptions:
            self.db.insert_assumption(assumption.model_dump())
        for ev in self.result.evidence:
            self.db.insert_evidence(ev.model_dump())
        for v in self.result.verifications:
            self.db.insert_verification(v.model_dump())
        for gap in self.result.gaps:
            self.db.insert_gap(gap.model_dump())

        for a in self.result.assumptions:
            self.db.insert_edge(a.claim_id, a.id, "GENERATES")
        for v in self.result.verifications:
            self.db.insert_edge(v.assumption_id, v.id, "VERIFIES")
            for ev_id in v.evidence_used:
                self.db.insert_edge(ev_id, v.id, "SUPPORTS")
        for g in self.result.gaps:
            self.db.insert_edge(g.assumption_id, g.id, "IDENTIFIES")

    def close(self) -> None:
        if self.db:
            self.db.close()

File: asf-tui/asf/analyzer/analyzer.go
package analyzer

import (
	"asf-tui/asf/assumption"
	"asf-tui/asf/confidence"
	"asf-tui/asf/evidence"
	"asf-tui/asf/extraction"
	"asf-tui/asf/gaps"
	"asf-tui/asf/graph"
	"asf-tui/asf/ingestion"
	"asf-tui/asf/models"
	"asf-tui/asf/verification"
)

type Analyzer struct {
	Pipeline           *ingestion.Pipeline
	ClaimExtractor     *extraction.ClaimExtractor
	AssumptionEngine   *assumption.Engine
	EvidenceLoader     *evidence.Loader
	EvidenceMapper     *evidence.Mapper
	VerificationEngine *verification.Engine
	ConfidenceEngine   *confidence.Engine
	GapEngine          *gaps.Engine
	GraphModel         *graph.Model
}

func New() *Analyzer {
	return &Analyzer{
		Pipeline:           ingestion.NewPipeline(),
		ClaimExtractor:     extraction.NewClaimExtractor(),
		AssumptionEngine:   assumption.NewEngine(),
		EvidenceLoader:     evidence.NewLoader(),
		EvidenceMapper:     evidence.NewMapper(),
		VerificationEngine: verification.NewEngine(),
		ConfidenceEngine:   confidence.NewEngine(),
		GapEngine:          gaps.NewEngine(),
		GraphModel:         graph.NewModel(),
	}
}

type AnalyzeResult struct {
	Result models.AnalysisResult
	Graph  graph.GraphData
}

func (a *Analyzer) Analyze(documentPaths []string, evidencePaths []string) (*AnalyzeResult, error) {
	var result models.AnalysisResult

	claims := a.processDocuments(documentPaths)
	assumptions := a.AssumptionEngine.ConvertMany(claims)

	var evidenceRecords []models.Evidence
	if len(evidencePaths) > 0 {
		evidenceRecords = a.loadEvidence(evidencePaths)
	}

	var verifications []models.Verification
	for _, assumption := range assumptions {
		matchingEvidence := a.findMatchingEvidence(assumption, evidenceRecords)
		verification := a.VerificationEngine.Verify(assumption, matchingEvidence)
		verification.Confidence = a.ConfidenceEngine.ComputeVerificationConfidence(verification, matchingEvidence)
		verifications = append(verifications, verification)

		assumption.Confidence = a.ConfidenceEngine.ComputeAssumptionConfidence([]models.Verification{verification})
		switch verification.Result {
		case models.VerificationResultVERIFIED:
			assumption.VerificationStatus = models.VerificationStatusVERIFIED
		case models.VerificationResultCONTRADICTED:
			assumption.VerificationStatus = models.VerificationStatusCONTRADICTED
		case models.VerificationResultPARTIALLY_VERIFIED:
			assumption.VerificationStatus = models.VerificationStatusIN_REVIEW
		}

		for i := range assumptions {
			if assumptions[i].ID == assumption.ID {
				assumptions[i] = assumption
				break
			}
		}
	}

	gapsResult := a.GapEngine.GenerateGaps(assumptions, verifications)

	result = models.AnalysisResult{
		Claims:        claims,
		Assumptions:   assumptions,
		Evidence:      evidenceRecords,
		Verifications: verifications,
		Gaps:          gapsResult,
	}

	a.GraphModel.Build(result)
	graphData := a.GraphModel.ExportData()

	return &AnalyzeResult{
		Result: result,
		Graph:  graphData,
	}, nil
}

func (a *Analyzer) processDocuments(paths []string) []models.Claim {
	var allClaims []models.Claim
	for _, path := range paths {
		text, err := a.Pipeline.ParseText(path)
		if err != nil {
			continue
		}
		meta := a.Pipeline.DetectType(path)
		_ = meta
		filename := path
		for i := len(path) - 1; i >= 0; i-- {
			if path[i] == '/' || path[i] == '\\' {
				filename = path[i+1:]
				break
			}
		}
		claims := a.ClaimExtractor.Extract(text, filename, path)
		allClaims = append(allClaims, claims...)
	}
	return allClaims
}

func (a *Analyzer) loadEvidence(paths []string) []models.Evidence {
	var evidenceList []models.Evidence
	for _, path := range paths {
		ev, err := a.EvidenceLoader.Load(path)
		if err != nil {
			continue
		}
		evidenceList = append(evidenceList, *ev)
	}
	return evidenceList
}

func (a *Analyzer) findMatchingEvidence(assumption models.Assumption, evidenceList []models.Evidence) []models.Evidence {
	compatibleTypes := a.EvidenceMapper.GetCompatibleSourceTypes(assumption.AssumptionType)
	if len(compatibleTypes) == 0 {
		return evidenceList
	}
	var matched []models.Evidence
	for _, ev := range evidenceList {
		for _, ct := range compatibleTypes {
			if ev.SourceType == ct {
				matched = append(matched, ev)
				break
			}
		}
	}
	if len(matched) == 0 {
		return evidenceList
	}
	return matched
}
